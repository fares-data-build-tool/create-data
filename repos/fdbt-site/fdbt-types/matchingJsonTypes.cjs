"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ExpiryUnit = exports.CarnetExpiryUnit = void 0;
var CarnetExpiryUnit;
(function (CarnetExpiryUnit) {
    CarnetExpiryUnit["HOUR"] = "hour";
    CarnetExpiryUnit["DAY"] = "day";
    CarnetExpiryUnit["WEEK"] = "week";
    CarnetExpiryUnit["MONTH"] = "month";
    CarnetExpiryUnit["YEAR"] = "year";
    CarnetExpiryUnit["NO_EXPIRY"] = "no expiry";
})(CarnetExpiryUnit = exports.CarnetExpiryUnit || (exports.CarnetExpiryUnit = {}));
var ExpiryUnit;
(function (ExpiryUnit) {
    ExpiryUnit["HOUR"] = "hour";
    ExpiryUnit["DAY"] = "day";
    ExpiryUnit["WEEK"] = "week";
    ExpiryUnit["MONTH"] = "month";
    ExpiryUnit["YEAR"] = "year";
})(ExpiryUnit = exports.ExpiryUnit || (exports.ExpiryUnit = {}));
