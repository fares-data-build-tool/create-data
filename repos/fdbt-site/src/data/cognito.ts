import { CognitoIdentityServiceProvider } from 'aws-sdk';
import awsParamStore from 'aws-param-store';
import crypto from 'crypto';

const clientId = process.env.FDBT_USER_POOL_CLIENT_ID as string;
const userPoolId = process.env.FDBT_USER_POOL_ID as string;
let clientSecret: string | undefined;

const getCognitoClient = (): CognitoIdentityServiceProvider =>
    new CognitoIdentityServiceProvider({ region: 'eu-west-2' });

const cognito = getCognitoClient();

const calculateSecretHash = (username: string): string => {
    if (!clientSecret) {
        clientSecret = awsParamStore.getParameterSync('fdbt-cognito-client-secret', { region: 'eu-west-2' }).Value;
    }

    return crypto
        .createHmac('SHA256', clientSecret as string)
        .update(username + clientId)
        .digest('base64');
};

export const initiateAuth = async (
    username: string,
    password: string,
): Promise<CognitoIdentityServiceProvider.AdminInitiateAuthResponse> => {
    console.info('initiating auth...');

    const params: CognitoIdentityServiceProvider.AdminInitiateAuthRequest = {
        AuthFlow: 'ADMIN_USER_PASSWORD_AUTH',
        ClientId: clientId,
        UserPoolId: userPoolId,
        AuthParameters: {
            USERNAME: username,
            PASSWORD: password,
            SECRET_HASH: calculateSecretHash(username),
        },
    };

    try {
        const response = await cognito.adminInitiateAuth(params).promise();

        return response;
    } catch (error) {
        throw new Error(`Failed to authenticate user: ${error.stack}`);
    }
};

export const initiateRefreshAuth = async (
    username: string,
    refreshToken: string,
): Promise<CognitoIdentityServiceProvider.AdminInitiateAuthResponse> => {
    console.info('initiating refresh auth...');

    const params: CognitoIdentityServiceProvider.AdminInitiateAuthRequest = {
        AuthFlow: 'REFRESH_TOKEN_AUTH',
        ClientId: clientId,
        UserPoolId: userPoolId,
        AuthParameters: {
            REFRESH_TOKEN: refreshToken,
            SECRET_HASH: calculateSecretHash(username),
        },
    };

    try {
        const response = await cognito.adminInitiateAuth(params).promise();

        return response;
    } catch (error) {
        throw new Error(`Failed to refresh user session: ${error.stack}`);
    }
};

export const respondToNewPasswordChallenge = async (
    username: string,
    password: string,
    session: string,
): Promise<void> => {
    console.info('new password challenge initiated...');

    const params: CognitoIdentityServiceProvider.AdminRespondToAuthChallengeRequest = {
        ChallengeName: 'NEW_PASSWORD_REQUIRED',
        ClientId: clientId,
        UserPoolId: userPoolId,
        ChallengeResponses: {
            USERNAME: username,
            NEW_PASSWORD: password,
            SECRET_HASH: calculateSecretHash(username),
        },
        Session: session,
    };

    try {
        await cognito.adminRespondToAuthChallenge(params).promise();
    } catch (error) {
        throw new Error(`Failed to respond to password challenge: ${error.stack}`);
    }
};

export const globalSignOut = async (username: string): Promise<void> => {
    console.info('performing global sign out...');

    const params: CognitoIdentityServiceProvider.AdminUserGlobalSignOutRequest = {
        Username: username,
        UserPoolId: userPoolId,
    };

    try {
        await cognito.adminUserGlobalSignOut(params).promise();
    } catch (error) {
        throw new Error(`Failed to perform global sign out: ${error.stack}`);
    }
};

export const updateUserAttributes = async (
    username: string,
    attributes: { Name: string; Value: string }[],
): Promise<void> => {
    console.info('updating user attributes...');

    const params: CognitoIdentityServiceProvider.AdminUpdateUserAttributesRequest = {
        UserAttributes: attributes,
        UserPoolId: userPoolId,
        Username: username,
    };

    try {
        await cognito.adminUpdateUserAttributes(params).promise();
    } catch (error) {
        throw new Error(`Failed to update user attributes: ${error.stack}`);
    }
};

export const forgotPassword = async (username: string): Promise<void> => {
    console.info('start forgot password...');

    const params: CognitoIdentityServiceProvider.ForgotPasswordRequest = {
        ClientId: clientId,
        Username: username,
        SecretHash: calculateSecretHash(username),
    };

    try {
        await cognito.forgotPassword(params).promise();
    } catch (error) {
        throw new Error(`Failed to perform forgot password request: ${error.stack}`);
    }
};

export const confirmForgotPassword = async (
    username: string,
    confirmationCode: string,
    password: string,
): Promise<void> => {
    console.info('confirm forgot password...');

    const params: CognitoIdentityServiceProvider.ConfirmForgotPasswordRequest = {
        ClientId: clientId,
        Username: username,
        ConfirmationCode: confirmationCode,
        Password: password,
        SecretHash: calculateSecretHash(username),
    };

    try {
        await cognito.confirmForgotPassword(params).promise();
    } catch (error) {
        if (error?.code === 'ExpiredCodeException') {
            throw new Error('ExpiredCodeException');
        }
        throw new Error(`Failed to confirm forgotten password: ${error.stack}`);
    }
};

export const updateUserPassword = async (newPassword: string, username: string): Promise<void> => {
    console.info('updating user password...');

    const params: CognitoIdentityServiceProvider.AdminSetUserPasswordRequest = {
        Password: newPassword,
        Permanent: true,
        Username: username,
        UserPoolId: userPoolId,
    };

    try {
        await cognito.adminSetUserPassword(params).promise();
    } catch (error) {
        throw new Error(`Failed to update user password: ${error.stack}`);
    }
};
