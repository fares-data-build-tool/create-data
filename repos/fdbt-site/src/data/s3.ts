import AWS from 'aws-sdk';

export interface FareStage {
    stageName: string;
    prices: {
        price: string;
        fareZones: string[];
    }[];
}

export interface UserFareStages {
    fareStages: FareStage[];
}

export interface UserFareZone {
    FareZoneName: string;
    NaptanCodes: string;
    AtcoCodes: string;
}

const getS3Client = (): AWS.S3 => {
    let options = {};

    if (process.env.NODE_ENV === 'development') {
        options = {
            s3ForcePathStyle: true,
            accessKeyId: 'S3RVER',
            secretAccessKey: 'S3RVER',
            endpoint: new AWS.Endpoint('http://localhost:4572'),
        };
    }

    return new AWS.S3(options);
};

const s3 = getS3Client();

export const getUserFareStages = async (uuid: string): Promise<UserFareStages> => {
    if (!process.env.USER_DATA_BUCKET_NAME) {
        throw new Error('Environment variable for validated bucket not set');
    }

    const params = {
        Bucket: process.env.USER_DATA_BUCKET_NAME,
        Key: `${uuid}.json`,
    };
    const response = await s3.getObject(params).promise();
    const dataAsString = response.Body?.toString('utf-8') ?? '';

    return JSON.parse(dataAsString);
};

export const putStringInS3 = async (
    bucketName: string,
    key: string,
    text: string,
    contentType: string,
): Promise<void> => {
    const request: AWS.S3.Types.PutObjectRequest = {
        Bucket: bucketName,
        Key: key,
        Body: Buffer.from(text, 'binary'),
        ContentType: contentType,
    };

    await s3.putObject(request).promise();
};

export const putDataInS3 = async (
    data: UserFareZone[] | UserFareStages | string,
    key: string,
    processed: boolean,
): Promise<void> => {
    let contentType = '';
    let bucketName = '';

    if (!process.env.USER_DATA_BUCKET_NAME || !process.env.RAW_USER_DATA_BUCKET_NAME) {
        throw new Error('Bucket name environment variables not set.');
    }

    if (processed) {
        bucketName = process.env.USER_DATA_BUCKET_NAME;
        contentType = 'application/json; charset=utf-8';
    } else {
        bucketName = process.env.RAW_USER_DATA_BUCKET_NAME;
        contentType = 'text/csv; charset=utf-8';
    }

    await putStringInS3(bucketName, key, JSON.stringify(data), contentType);
};
