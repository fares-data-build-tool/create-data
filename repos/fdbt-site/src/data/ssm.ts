import { GetParameterCommand, SSMClient } from '@aws-sdk/client-ssm';

const ssm = new SSMClient({ region: 'eu-west-2' });

export const getSsmValue = async (name: string, withDecryption?: boolean): Promise<string> => {
    const input = {
        Name: name,
        WithDecryption: withDecryption ? withDecryption : true,
    };
    const value = (await ssm.send(new GetParameterCommand(input))).Parameter?.Value;

    if (!value) {
        throw new Error(`Parameter ${name} was not found`);
    }

    return value;
};
