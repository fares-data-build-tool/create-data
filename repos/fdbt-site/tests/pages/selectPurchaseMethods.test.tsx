import * as React from 'react';
import { shallow } from 'enzyme';
import { getMockContext } from '../testData/mockData';
import { getSalesOfferPackagesByNocCode } from '../../src/data/auroradb';
import {
    FARE_TYPE_ATTRIBUTE,
    TYPE_OF_CAP_ATTRIBUTE,
    PRICING_PER_DISTANCE_ATTRIBUTE,
    MULTIPLE_PRODUCT_ATTRIBUTE,
    OPERATOR_ATTRIBUTE,
} from '../../src/constants/attributes';
import SelectPurchaseMethods, { getServerSideProps, PurchaseMethodsProps } from '../../src/pages/selectPurchaseMethods';
import { FromDb, SalesOfferPackage } from '../../src/interfaces/matchingJsonTypes';

jest.mock('../../src/data/auroradb');

const defaultSalesOfferPackageOne: FromDb<SalesOfferPackage> = {
    id: 1,
    name: 'Onboard (cash)',
    description: '',
    purchaseLocations: ['onBoard'],
    paymentMethods: ['cash'],
    ticketFormats: ['paperTicket'],
    isCapped: false,
};

const defaultSalesOfferPackageTwo: FromDb<SalesOfferPackage> = {
    id: 2,
    name: 'Onboard (contactless)',
    description: '',
    purchaseLocations: ['onBoard'],
    paymentMethods: ['contactlessPaymentCard'],
    ticketFormats: ['paperTicket'],
    isCapped: false,
};

const defaultSalesOfferPackageThree: FromDb<SalesOfferPackage> = {
    id: 3,
    name: 'Online (smart card)',
    description: '',
    purchaseLocations: ['online'],
    paymentMethods: ['directDebit', 'creditCard', 'debitCard'],
    ticketFormats: ['smartCard'],
    isCapped: false,
};

const defaultSalesOfferPackageFour: FromDb<SalesOfferPackage> = {
    id: 4,
    name: 'Mobile App',
    description: '',
    purchaseLocations: ['mobileDevice'],
    paymentMethods: ['debitCard', 'creditCard', 'mobilePhone', 'directDebit'],
    ticketFormats: ['mobileApp'],
    isCapped: false,
};

const defaultSalesOfferPackageFive: FromDb<SalesOfferPackage> = {
    id: 4,
    name: 'Capped Purchase Method Mobile',
    description: '',
    purchaseLocations: ['mobileDevice'],
    paymentMethods: ['debitCard', 'creditCard'],
    ticketFormats: ['mobileApp'],
    isCapped: true,
};

const defaultSalesOfferPackageSix: FromDb<SalesOfferPackage> = {
    id: 4,
    name: 'Mobile App',
    description: '',
    purchaseLocations: ['onBoard'],
    paymentMethods: ['debitCard', 'creditCard', 'mobilePhone'],
    ticketFormats: ['smartCard'],
    isCapped: false,
};

describe('pages', () => {
    const selectSalesOfferPackagePropsInfoNoError: PurchaseMethodsProps = {
        purchaseMethodsList: [
            defaultSalesOfferPackageOne,
            defaultSalesOfferPackageTwo,
            defaultSalesOfferPackageThree,
            defaultSalesOfferPackageFour,
        ],
        errors: [],
        products: [],
        csrfToken: '',
        backHref: '',
        isCapped: false,
    };

    const selectCappedSalesOfferPackagePropsInfoNoError: PurchaseMethodsProps = {
        purchaseMethodsList: [defaultSalesOfferPackageFive, defaultSalesOfferPackageSix],
        errors: [],
        products: [],
        csrfToken: '',
        backHref: '',
        isCapped: true,
    };

    const selectSalesOfferPackagePropsInfoWithError: PurchaseMethodsProps = {
        products: [],
        purchaseMethodsList: [
            defaultSalesOfferPackageOne,
            defaultSalesOfferPackageTwo,
            defaultSalesOfferPackageThree,
            defaultSalesOfferPackageFour,
        ],
        errors: [{ errorMessage: 'Choose at least one service from the options', id: 'sales-offer-package-error' }],
        csrfToken: '',
        backHref: '',
        isCapped: false,
    };

    const selectCappedSalesOfferPackagePropsInfoWithError: PurchaseMethodsProps = {
        products: [],
        purchaseMethodsList: [defaultSalesOfferPackageFive, defaultSalesOfferPackageSix],
        errors: [{ errorMessage: 'Choose at least one service from the options', id: 'sales-offer-package-error' }],
        csrfToken: '',
        backHref: '',
        isCapped: false,
    };

    describe('selectPurchaseMethods', () => {
        it('should render correctly', () => {
            const tree = shallow(
                <SelectPurchaseMethods
                    purchaseMethodsList={selectSalesOfferPackagePropsInfoNoError.purchaseMethodsList}
                    products={[
                        {
                            productName: 'Great Product',
                            productPrice: '22',
                        },
                    ]}
                    errors={selectSalesOfferPackagePropsInfoNoError.errors}
                    csrfToken=""
                    backHref=""
                    isCapped={false}
                />,
            );
            expect(tree).toMatchSnapshot();
        });

        it('should render correctly for capped ticket', () => {
            const tree = shallow(
                <SelectPurchaseMethods
                    purchaseMethodsList={selectCappedSalesOfferPackagePropsInfoNoError.purchaseMethodsList}
                    products={[
                        {
                            productName: 'Great Product',
                            productPrice: '22',
                        },
                    ]}
                    errors={selectCappedSalesOfferPackagePropsInfoNoError.errors}
                    csrfToken=""
                    backHref=""
                    isCapped={true}
                />,
            );
            expect(tree).toMatchSnapshot();
        });

        it('should render correctly with no purchase methods', () => {
            const tree = shallow(
                <SelectPurchaseMethods
                    purchaseMethodsList={[]}
                    products={[
                        {
                            productName: 'Great Product',
                            productPrice: '22',
                        },
                    ]}
                    errors={selectSalesOfferPackagePropsInfoNoError.errors}
                    csrfToken=""
                    backHref=""
                    isCapped={false}
                />,
            );
            expect(tree).toMatchSnapshot();
        });

        it('should render correctly with no purchase methods for capped ticket', () => {
            const tree = shallow(
                <SelectPurchaseMethods
                    purchaseMethodsList={[]}
                    products={[
                        {
                            productName: 'Great Product',
                            productPrice: '22',
                        },
                    ]}
                    errors={selectCappedSalesOfferPackagePropsInfoNoError.errors}
                    csrfToken=""
                    backHref=""
                    isCapped={true}
                />,
            );
            expect(tree).toMatchSnapshot();
        });

        it('should render an error when an error message is passed through to props', () => {
            const tree = shallow(
                <SelectPurchaseMethods
                    purchaseMethodsList={selectSalesOfferPackagePropsInfoWithError.purchaseMethodsList}
                    products={[]}
                    errors={selectSalesOfferPackagePropsInfoWithError.errors}
                    csrfToken=""
                    backHref=""
                    isCapped={false}
                />,
            );
            expect(tree).toMatchSnapshot();
        });

        it('should render an error when an error message is passed through to props for capped ticket', () => {
            const tree = shallow(
                <SelectPurchaseMethods
                    purchaseMethodsList={selectCappedSalesOfferPackagePropsInfoWithError.purchaseMethodsList}
                    products={[]}
                    errors={selectCappedSalesOfferPackagePropsInfoWithError.errors}
                    csrfToken=""
                    backHref=""
                    isCapped={true}
                />,
            );
            expect(tree).toMatchSnapshot();
        });

        describe('getServerSideProps', () => {
            it.each([
                ['single', '', false],
                ['flatFare', '', false],
                ['period', ' geoZone', false],
                ['period', ' mulitpleServices', true],
                ['multiOperator', ' geoZone', true],
                ['multiOperator', ' multipleServices', false],
            ])(
                'should return expected props when the page is visited by a user with stored sales offer packages on the %s%s ticket flow',
                async (fareType, _ticketType, multipleProducts) => {
                    const mockSalesOfferPackages: SalesOfferPackage[] = [
                        {
                            id: 1,
                            name: 'IW Bus Co - Package 1',
                            description: 'On bus - Cash - Paper Ticket',
                            purchaseLocations: ['On bus'],
                            paymentMethods: ['Cash'],
                            ticketFormats: ['Paper ticket'],
                            isCapped: false,
                        },
                        {
                            id: 2,
                            name: 'IW Bus Co - Package 2',
                            description: 'Online App - Card via App - eTicket',
                            purchaseLocations: ['Online App'],
                            paymentMethods: ['Card via App'],
                            ticketFormats: ['eTicket'],
                            isCapped: false,
                        },
                        {
                            id: 3,
                            name: 'IW Bus Co - Package 3',
                            description: 'Ticket Machine - Card - Pass',
                            purchaseLocations: ['Ticket Machine'],
                            paymentMethods: ['Card'],
                            ticketFormats: ['Pass'],
                            isCapped: false,
                        },
                    ];
                    (getSalesOfferPackagesByNocCode as jest.Mock).mockImplementation(() => mockSalesOfferPackages);
                    const ctx = getMockContext({
                        session: {
                            [FARE_TYPE_ATTRIBUTE]: {
                                fareType,
                            },
                            ...(!multipleProducts && { [MULTIPLE_PRODUCT_ATTRIBUTE]: null }),
                        },
                    });
                    const result = await getServerSideProps(ctx);
                    const expectedSalesOfferPackageList: SalesOfferPackage[] = mockSalesOfferPackages.map(
                        (mockSalesOfferPackage) => {
                            return {
                                ...mockSalesOfferPackage,
                            };
                        },
                    );
                    const expectedProductNamesLength = !multipleProducts
                        ? 1
                        : ctx.req.session[MULTIPLE_PRODUCT_ATTRIBUTE].products.length;
                    expect(result.props.errors.length).toBe(0);
                    expect(result.props.products.length).toBe(expectedProductNamesLength);
                    expect(result.props.purchaseMethodsList).toEqual(expectedSalesOfferPackageList);
                },
            );

            it('should return expected props when the page is visited by a user with stored sales offer packages on the capped ticket', async () => {
                const mockSalesOfferPackages: SalesOfferPackage[] = [
                    {
                        id: 1,
                        name: 'Capped Ticket',
                        description: 'On bus - CreditCard - mobile',
                        purchaseLocations: ['On Board'],
                        paymentMethods: ['CreditCard'],
                        ticketFormats: ['Mobile'],
                        isCapped: true,
                    },
                ];
                (getSalesOfferPackagesByNocCode as jest.Mock).mockImplementation(() => mockSalesOfferPackages);

                const ctx = getMockContext({
                    session: {
                        [FARE_TYPE_ATTRIBUTE]: { fareType: 'capped' },
                        [TYPE_OF_CAP_ATTRIBUTE]: { typeOfCap: 'byDistance' },
                        [PRICING_PER_DISTANCE_ATTRIBUTE]: {
                            productName: 'product name',
                            maximumPrice: '',
                            minimumPrice: '',
                            capPricing: [],
                        },
                    },
                });

                const result = await getServerSideProps(ctx);
                const expectedSalesOfferPackageList: SalesOfferPackage[] = mockSalesOfferPackages.map(
                    (mockSalesOfferPackage) => {
                        return {
                            ...mockSalesOfferPackage,
                        };
                    },
                );

                expect(result.props.errors.length).toBe(0);
                expect(result.props.products.length).toBe(1);
                expect(result.props.purchaseMethodsList).toEqual(expectedSalesOfferPackageList);
            });

            it('should throw an error when necessary nocCode is invalid, when the user is not a scheme operator', async () => {
                const ctx = getMockContext({
                    session: { [OPERATOR_ATTRIBUTE]: undefined },
                    body: null,
                    uuid: {},
                    mockWriteHeadFn: jest.fn(),
                    mockEndFn: jest.fn(),
                    isLoggedin: false,
                });
                await expect(getServerSideProps(ctx)).rejects.toThrow('invalid NOC set');
            });
        });
    });
});
