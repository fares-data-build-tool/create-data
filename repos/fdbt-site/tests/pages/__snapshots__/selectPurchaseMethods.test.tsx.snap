// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`pages selectPurchaseMethods should render an error when an error message is passed through to props 1`] = `
<FullColumnLayout
  description="Purchase Methods selection page of the Create Fares Data Service"
  title="Select Purchase Methods - Create Fares Data Service"
>
  <CsrfForm
    action="/api/selectSalesOfferPackage"
    csrfToken=""
    method="post"
  >
    <ErrorSummary
      errors={
        Array [
          Object {
            "errorMessage": "Choose at least one service from the options",
            "id": "sales-offer-package-error",
          },
        ]
      }
    />
    <h1
      className="govuk-heading-l"
      id="select-sales-offer-package-page-heading"
    >
      How are your tickets sold?
    </h1>
    <div
      className="govuk-warning-text"
    >
      <span
        aria-hidden="true"
        className="govuk-warning-text__icon govuk-!-margin-top-1"
      >
        !
      </span>
      <strong
        className="govuk-warning-text__text"
      >
        <span
          className="govuk-warning-text__assistive"
        >
          Warning
        </span>
        You can create new purchase methods in your
         
        <a
          className="govuk-link"
          href="/viewPurchaseMethods"
        >
          operator settings.
        </a>
         
        <br />
        Don't worry you can navigate back to this page when you are finished.
      </strong>
    </div>
    <input
      className="govuk-button"
      id="continue-button"
      type="submit"
      value="Continue"
    />
    <a
      className="govuk-button govuk-button--secondary create-new-sop-button"
      data-module="govuk-button"
      draggable="false"
      href="/viewPurchaseMethods"
      id="create-new-button"
      role="button"
    >
      Create new
    </a>
  </CsrfForm>
</FullColumnLayout>
`;

exports[`pages selectPurchaseMethods should render correctly 1`] = `
<FullColumnLayout
  description="Purchase Methods selection page of the Create Fares Data Service"
  title="Select Purchase Methods - Create Fares Data Service"
>
  <CsrfForm
    action="/api/selectSalesOfferPackage"
    csrfToken=""
    method="post"
  >
    <ErrorSummary
      errors={Array []}
    />
    <h1
      className="govuk-heading-l"
      id="select-sales-offer-package-page-heading"
    >
      How are your tickets sold?
    </h1>
    <div
      className="govuk-warning-text"
    >
      <span
        aria-hidden="true"
        className="govuk-warning-text__icon govuk-!-margin-top-1"
      >
        !
      </span>
      <strong
        className="govuk-warning-text__text"
      >
        <span
          className="govuk-warning-text__assistive"
        >
          Warning
        </span>
        You can create new purchase methods in your
         
        <a
          className="govuk-link"
          href="/viewPurchaseMethods"
        >
          operator settings.
        </a>
         
        <br />
        Don't worry you can navigate back to this page when you are finished.
      </strong>
    </div>
    <div
      className="sop-option"
      key="Great Product"
    >
      <FormGroupWrapper
        errorIds={
          Array [
            "product-GreatProduct-checkbox-0",
          ]
        }
        errors={Array []}
        hideErrorBar={false}
      >
        <fieldset
          className="govuk-fieldset"
        >
          <h1
            className="govuk-heading-m govuk-!-font-weight-regular govuk-!-padding-top-6 govuk-!-padding-bottom-2"
          >
            How is 
            <span
              className="govuk-!-font-weight-bold"
            >
              Great Product
            </span>
             sold?
          </h1>
          <FormElementWrapper
            errorClass=""
            errorId="product-GreatProduct-checkbox-0"
            errors={Array []}
          >
            <div
              className="govuk-checkboxes"
            >
              <div
                className="card-row"
              >
                <div
                  className="card"
                  key="checkbox-item-Onboard (cash)"
                >
                  <div
                    className="card__body card_align"
                  >
                    <div
                      className="govuk-checkboxes__item card__selector"
                    >
                      <input
                        className="govuk-checkboxes__input"
                        defaultChecked={false}
                        id="product-GreatProduct-checkbox-0"
                        name="^0^0-product-Great Product"
                        onChange={[Function]}
                        type="checkbox"
                        value="{\\"id\\":1,\\"name\\":\\"Onboard (cash)\\",\\"description\\":\\"\\",\\"purchaseLocations\\":[\\"onBoard\\"],\\"paymentMethods\\":[\\"cash\\"],\\"ticketFormats\\":[\\"paperTicket\\"]}"
                      />
                      <label
                        className="govuk-label govuk-checkboxes__label"
                        htmlFor="product-GreatProduct-checkbox-0"
                      />
                    </div>
                    <PurchaseMethodCardBody
                      entity={
                        Object {
                          "description": "",
                          "id": 1,
                          "name": "Onboard (cash)",
                          "paymentMethods": Array [
                            "cash",
                          ],
                          "purchaseLocations": Array [
                            "onBoard",
                          ],
                          "ticketFormats": Array [
                            "paperTicket",
                          ],
                        }
                      }
                    />
                  </div>
                </div>
                <div
                  className="card"
                  key="checkbox-item-Onboard (contactless)"
                >
                  <div
                    className="card__body card_align"
                  >
                    <div
                      className="govuk-checkboxes__item card__selector"
                    >
                      <input
                        className="govuk-checkboxes__input"
                        defaultChecked={false}
                        id="product-GreatProduct-checkbox-1"
                        name="^0^1-product-Great Product"
                        onChange={[Function]}
                        type="checkbox"
                        value="{\\"id\\":2,\\"name\\":\\"Onboard (contactless)\\",\\"description\\":\\"\\",\\"purchaseLocations\\":[\\"onBoard\\"],\\"paymentMethods\\":[\\"contactlessPaymentCard\\"],\\"ticketFormats\\":[\\"paperTicket\\"]}"
                      />
                      <label
                        className="govuk-label govuk-checkboxes__label"
                        htmlFor="product-GreatProduct-checkbox-1"
                      />
                    </div>
                    <PurchaseMethodCardBody
                      entity={
                        Object {
                          "description": "",
                          "id": 2,
                          "name": "Onboard (contactless)",
                          "paymentMethods": Array [
                            "contactlessPaymentCard",
                          ],
                          "purchaseLocations": Array [
                            "onBoard",
                          ],
                          "ticketFormats": Array [
                            "paperTicket",
                          ],
                        }
                      }
                    />
                  </div>
                </div>
                <div
                  className="card"
                  key="checkbox-item-Online (smart card)"
                >
                  <div
                    className="card__body card_align"
                  >
                    <div
                      className="govuk-checkboxes__item card__selector"
                    >
                      <input
                        className="govuk-checkboxes__input"
                        defaultChecked={false}
                        id="product-GreatProduct-checkbox-2"
                        name="^0^2-product-Great Product"
                        onChange={[Function]}
                        type="checkbox"
                        value="{\\"id\\":3,\\"name\\":\\"Online (smart card)\\",\\"description\\":\\"\\",\\"purchaseLocations\\":[\\"online\\"],\\"paymentMethods\\":[\\"directDebit\\",\\"creditCard\\",\\"debitCard\\"],\\"ticketFormats\\":[\\"smartCard\\"]}"
                      />
                      <label
                        className="govuk-label govuk-checkboxes__label"
                        htmlFor="product-GreatProduct-checkbox-2"
                      />
                    </div>
                    <PurchaseMethodCardBody
                      entity={
                        Object {
                          "description": "",
                          "id": 3,
                          "name": "Online (smart card)",
                          "paymentMethods": Array [
                            "directDebit",
                            "creditCard",
                            "debitCard",
                          ],
                          "purchaseLocations": Array [
                            "online",
                          ],
                          "ticketFormats": Array [
                            "smartCard",
                          ],
                        }
                      }
                    />
                  </div>
                </div>
                <div
                  className="card"
                  key="checkbox-item-Mobile App"
                >
                  <div
                    className="card__body card_align"
                  >
                    <div
                      className="govuk-checkboxes__item card__selector"
                    >
                      <input
                        className="govuk-checkboxes__input"
                        defaultChecked={false}
                        id="product-GreatProduct-checkbox-3"
                        name="^0^3-product-Great Product"
                        onChange={[Function]}
                        type="checkbox"
                        value="{\\"id\\":4,\\"name\\":\\"Mobile App\\",\\"description\\":\\"\\",\\"purchaseLocations\\":[\\"mobileDevice\\"],\\"paymentMethods\\":[\\"debitCard\\",\\"creditCard\\",\\"mobilePhone\\",\\"directDebit\\"],\\"ticketFormats\\":[\\"mobileApp\\"]}"
                      />
                      <label
                        className="govuk-label govuk-checkboxes__label"
                        htmlFor="product-GreatProduct-checkbox-3"
                      />
                    </div>
                    <PurchaseMethodCardBody
                      entity={
                        Object {
                          "description": "",
                          "id": 4,
                          "name": "Mobile App",
                          "paymentMethods": Array [
                            "debitCard",
                            "creditCard",
                            "mobilePhone",
                            "directDebit",
                          ],
                          "purchaseLocations": Array [
                            "mobileDevice",
                          ],
                          "ticketFormats": Array [
                            "mobileApp",
                          ],
                        }
                      }
                    />
                  </div>
                </div>
              </div>
              <input
                name="product-Great Product"
                type="hidden"
              />
            </div>
          </FormElementWrapper>
        </fieldset>
      </FormGroupWrapper>
    </div>
    <input
      className="govuk-button"
      id="continue-button"
      type="submit"
      value="Continue"
    />
    <a
      className="govuk-button govuk-button--secondary create-new-sop-button"
      data-module="govuk-button"
      draggable="false"
      href="/viewPurchaseMethods"
      id="create-new-button"
      role="button"
    >
      Create new
    </a>
  </CsrfForm>
</FullColumnLayout>
`;

exports[`pages selectPurchaseMethods should render correctly with no purchase methods 1`] = `
<FullColumnLayout
  description="Purchase Methods selection page of the Create Fares Data Service"
  title="Select Purchase Methods - Create Fares Data Service"
>
  <CsrfForm
    action="/api/selectSalesOfferPackage"
    csrfToken=""
    method="post"
  >
    <ErrorSummary
      errors={Array []}
    />
    <h1
      className="govuk-heading-l"
      id="select-sales-offer-package-page-heading"
    >
      How are your tickets sold?
    </h1>
    <div
      className="govuk-warning-text"
    >
      <span
        aria-hidden="true"
        className="govuk-warning-text__icon govuk-!-margin-top-1"
      >
        !
      </span>
      <strong
        className="govuk-warning-text__text"
      >
        <span
          className="govuk-warning-text__assistive"
        >
          Warning
        </span>
        You can create new purchase methods in your
         
        <a
          className="govuk-link"
          href="/viewPurchaseMethods"
        >
          operator settings.
        </a>
         
        <br />
        Don't worry you can navigate back to this page when you are finished.
      </strong>
    </div>
    <div
      className="sop-option"
      key="Great Product"
    >
      <FormGroupWrapper
        errorIds={
          Array [
            "product-GreatProduct-checkbox-0",
          ]
        }
        errors={Array []}
        hideErrorBar={false}
      >
        <fieldset
          className="govuk-fieldset"
        >
          <h1
            className="govuk-heading-m govuk-!-font-weight-regular govuk-!-padding-top-6 govuk-!-padding-bottom-2"
          >
            How is 
            <span
              className="govuk-!-font-weight-bold"
            >
              Great Product
            </span>
             sold?
          </h1>
          <FormElementWrapper
            errorClass=""
            errorId="product-GreatProduct-checkbox-0"
            errors={Array []}
          >
            <span
              className="govuk-body"
            >
              <i>
                You currently have no saved purchase methods
              </i>
            </span>
          </FormElementWrapper>
        </fieldset>
      </FormGroupWrapper>
    </div>
    <input
      className="govuk-button"
      id="continue-button"
      type="submit"
      value="Continue"
    />
    <a
      className="govuk-button govuk-button--secondary create-new-sop-button"
      data-module="govuk-button"
      draggable="false"
      href="/viewPurchaseMethods"
      id="create-new-button"
      role="button"
    >
      Create new
    </a>
  </CsrfForm>
</FullColumnLayout>
`;
